File: latest/week3/workshop7_m1/W07_M01_demo.py
"""W07-M01: Narrative Structure in Data Communication — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
from matplotlib.dates import DateFormatter
os.makedirs("output", exist_ok=True)

# ── 1. LOAD DATA ────────────────────────────────────────
nf = pd.read_csv("netflix.csv")
nf["date_added"] = pd.to_datetime(nf["date_added"].str.strip(), errors="coerce")
nf["year_added"] = nf["date_added"].dt.year

yearly = (nf.query("2015 <= year_added <= 2021")
    .groupby(["year_added", "type"]).size()
    .reset_index(name="n"))

# ── 2. EXPLORATORY vs EXPLANATORY ────────────────────────
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(16, 5))

# EXPLORATORY: raw EDA chart
for t, c, m in [("Movie", "#1565C0", "o"), ("TV Show", "#E53935", "s")]:
    sub = yearly.query("type == @t")
    ax1.plot(sub.year_added, sub.n, f"-{m}", color=c, lw=1, markersize=5, label=t)
ax1.legend(fontsize=8)
ax1.set_title("EXPLORATORY\n'Netflix Additions by Type (2015–2021)'\n"
    "(descriptive title, legend, no emphasis)", fontsize=10, fontweight="bold", color="#888")
ax1.set_xlabel("Year"); ax1.set_ylabel("Count")
ax1.spines["top"].set_visible(False); ax1.spines["right"].set_visible(False)

# EXPLANATORY: story-focused redesign
movies = yearly.query("type == 'Movie'")
tvshows = yearly.query("type == 'TV Show'")
ax2.plot(tvshows.year_added, tvshows.n, "-o", color="#DDDDDD", lw=0.8, markersize=4)
ax2.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2.5, markersize=6)
# Direct labels
ax2.text(2021.15, movies.iloc[-1]["n"], " Movies", fontsize=9, fontweight="bold",
    color="#1565C0", va="center")
ax2.text(2021.15, tvshows.iloc[-1]["n"], " TV Shows", fontsize=9,
    color="#BBBBBB", va="center")
ax2.set_xlim(2014.5, 2022.5)
ax2.set_title("EXPLANATORY\n'Movie additions peaked in 2019 and declined since'\n"
    "(declarative title, direct labels, grey+accent)", fontsize=10, fontweight="bold", color="#1565C0")
ax2.set_ylabel("Titles Added")
ax2.spines["top"].set_visible(False); ax2.spines["right"].set_visible(False)

fig.suptitle("The Fundamental Shift: From Exploration to Explanation",
    fontsize=13, fontweight="bold")
plt.tight_layout()
plt.savefig("output/exp_vs_expl.png", dpi=300); plt.close()

# ── 3. BIG IDEA CHART ───────────────────────────────────
peak = movies.loc[movies.n.idxmax()]
latest = movies.iloc[-1]
decline_pct = round((1 - latest["n"] / peak["n"]) * 100)

fig, ax = plt.subplots(figsize=(9, 5))
ax.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2.5, markersize=6)
# Peak reference line
ax.axhline(y=peak["n"], ls=":", color="#888", lw=0.5)
# Decline annotation with arrow
ax.annotate(f"{decline_pct}% decline\nsince peak",
    xy=(2020.5, (peak["n"] + latest["n"]) / 2),
    fontsize=11, fontweight="bold", color="#E53935", ha="center",
    bbox=dict(boxstyle="round,pad=0.3", facecolor="white", edgecolor="#E53935", alpha=0.9))
ax.annotate("", xy=(2020.5, latest["n"] + 10),
    xytext=(2020.5, peak["n"] - 10),
    arrowprops=dict(arrowstyle="<->", color="#E53935", lw=1.5))
ax.set_title(f"Movie additions declined {decline_pct}% from peak —\n"
    "the content pipeline is contracting",
    fontsize=12, fontweight="bold")
ax.set_ylabel("Movie Titles Added"); ax.set_xlabel("")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout()
plt.savefig("output/big_idea.png", dpi=300); plt.close()

# ── 4. THREE-SLIDE STORY ARC ────────────────────────────
fig, axes = plt.subplots(3, 1, figsize=(10, 14))

# Slide 1: Context (cumulative catalog growth)
cumulative = (nf.dropna(subset=["date_added"])
    .set_index("date_added").resample("ME").size().cumsum())
axes[0].fill_between(cumulative.index, cumulative.values, alpha=0.2, color="#1565C0")
axes[0].plot(cumulative.index, cumulative.values, color="#1565C0", lw=1.5)
axes[0].set_title("1. CONTEXT: Netflix built 8,800+ titles by 2021",
    fontsize=11, fontweight="bold", color="#2E7D32")
axes[0].set_ylabel("Cumulative Titles")
axes[0].xaxis.set_major_formatter(DateFormatter("%Y"))

# Slide 2: Complication (movie decline)
axes[1].plot(tvshows.year_added, tvshows.n, "-o", color="#DDDDDD", lw=0.8, markersize=4)
axes[1].plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2.5, markersize=6)
axes[1].text(2021.1, movies.iloc[-1]["n"], " Movies", fontsize=9,
    fontweight="bold", color="#1565C0", va="center")
axes[1].text(2021.1, tvshows.iloc[-1]["n"], " TV Shows", fontsize=8,
    color="#BBBBBB", va="center")
axes[1].set_xlim(2014.5, 2022.5)
axes[1].set_title("2. COMPLICATION: Movie additions peaked and declined",
    fontsize=11, fontweight="bold", color="#E65100")
axes[1].set_ylabel("Titles Added")

# Slide 3: Resolution (genre growth)
genres = nf["listed_in"].str.split(",").explode().str.strip()
g17 = nf.query("year_added == 2017")["listed_in"].str.split(",").explode().str.strip().value_counts()
g21 = nf.query("year_added == 2021")["listed_in"].str.split(",").explode().str.strip().value_counts()
growth = ((g21 - g17) / g17.clip(lower=1) * 100).dropna().sort_values().tail(5)
axes[2].barh(growth.index, growth.values, color="#2E7D32", height=0.5)
for i, (genre, val) in enumerate(growth.items()):
    axes[2].text(val + 2, i, f"+{val:.0f}%", va="center", fontsize=8,
        fontweight="bold", color="#2E7D32")
axes[2].set_title("3. RESOLUTION: International & niche genres are the growth engines",
    fontsize=11, fontweight="bold", color="#7B1FA2")
axes[2].set_xlabel("Growth % (2017 → 2021)")

for ax in axes:
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
fig.suptitle("Three-Slide Story Arc: Context → Complication → Resolution",
    fontsize=13, fontweight="bold")
plt.tight_layout()
plt.savefig("output/story_arc_3slides.png", dpi=300); plt.close()

# ── 5. TITLE COMPARISON ─────────────────────────────────
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(14, 4.5))
for ax, title, subtitle in [
    (ax1, "Descriptive: 'Netflix Additions by Type'",
     "(What the chart shows — not what it means)"),
    (ax2, "Declarative: 'Movies peaked 2019; TV grew steadily'",
     "(The title IS the finding — the chart is the evidence)")]:
    for t, c in [("Movie", "#1565C0"), ("TV Show", "#E53935")]:
        sub = yearly.query("type == @t")
        ax.plot(sub.year_added, sub.n, "-o", color=c, lw=1.2, markersize=4, label=t)
    ax.legend(fontsize=6); ax.set_title(f"{title}\n{subtitle}",
        fontsize=9, fontweight="bold")
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
fig.suptitle("Chart Titles: Descriptive vs Declarative", fontweight="bold")
plt.tight_layout()
plt.savefig("output/title_comparison.png", dpi=300); plt.close()

print("\nAll W07-M01 Python outputs saved")
print("Figures: exp_vs_expl, big_idea, story_arc_3slides, title_comparison")

File: latest/week3/workshop7_m1/W07_M01_demo.R
# ============================================================
# Workshop 7 — Module 1: Narrative Structure in Data Communication
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(patchwork); dir.create("output", showWarnings = FALSE)

# This module demonstrates the TRANSITION from exploratory (W04-W06)
# to explanatory (W07) using the Netflix dataset.

# ── 1. LOAD DATA ────────────────────────────────────────
nf <- read_csv("netflix.csv") |>
  mutate(
    date_added = mdy(str_trim(date_added)),
    year_added = year(date_added),
    primary_country = str_trim(str_extract(country, "^[^,]+")))

nf_genres <- nf |>
  separate_rows(listed_in, sep = ", ") |>
  mutate(genre = str_trim(listed_in))

# ── 2. EXPLORATORY VERSION (how you found the insight) ──
# Raw EDA chart — fine for the analyst, bad for stakeholders
yearly_type <- nf |>
  filter(!is.na(year_added), year_added >= 2015, year_added <= 2021) |>
  count(year_added, type)

p_exploratory <- ggplot(yearly_type, aes(x = year_added, y = n, color = type)) +
  geom_line(linewidth = 1) + geom_point(size = 2) +
  scale_color_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  theme_minimal() +
  labs(title = "Netflix: Yearly Additions by Type (2015–2021)",
       x = "Year", y = "Count", color = "Type")
ggsave("output/exploratory_version.png", p_exploratory, width = 8, height = 5, dpi = 300)

# ── 3. EXPLANATORY VERSION (how you tell the story) ─────
# Same data, redesigned for the audience
p_explanatory <- ggplot(yearly_type, aes(x = year_added, y = n, group = type)) +
  # Grey all, accent the story
  geom_line(color = "#DDDDDD", linewidth = 0.8) +
  geom_line(data = yearly_type |> filter(type == "Movie"),
    color = "#1565C0", linewidth = 2) +
  geom_point(data = yearly_type |> filter(type == "Movie"),
    color = "#1565C0", size = 3) +
  # Direct labels (no legend needed)
  geom_text(data = yearly_type |> filter(year_added == 2021),
    aes(label = type, color = type == "Movie"),
    hjust = -0.15, size = 3.5, fontface = "bold", show.legend = FALSE) +
  scale_color_manual(values = c("TRUE" = "#1565C0", "FALSE" = "#BBBBBB")) +
  # Declarative title = the finding
  theme_minimal(base_size = 11) +
  theme(legend.position = "none",
    plot.title = element_text(face = "bold", size = 13),
    plot.subtitle = element_text(color = "#666")) +
  coord_cartesian(xlim = c(2015, 2022.5)) +
  labs(title = "Netflix movie additions peaked in 2019 and have since declined",
       subtitle = "TV Show additions grew steadily, but movies — the larger category — are contracting",
       x = NULL, y = "Titles Added per Year",
       caption = "Source: Netflix dataset | UNYT Data Viz Course")
ggsave("output/explanatory_version.png", p_explanatory, width = 9, height = 5, dpi = 300)

# Side-by-side comparison
ggsave("output/exp_vs_expl.png",
  (p_exploratory + labs(title = "EXPLORATORY\n(descriptive title, legend, no story)")) |
  (p_explanatory + labs(title = "EXPLANATORY\n(declarative title, direct labels, focused)")),
  width = 16, height = 5, dpi = 300)

# ── 4. THE BIG IDEA EXERCISE ────────────────────────────
# Netflix Big Idea:
# "Netflix should accelerate original series investment because
#  movie additions have declined 40% since 2019 while TV Show
#  additions continue to grow."
#
# Structure: [Subject] should [action] because [evidence]

# Visualize the Big Idea as a single annotated chart
movie_peak <- yearly_type |> filter(type == "Movie") |> slice_max(n, n = 1)
movie_latest <- yearly_type |> filter(type == "Movie", year_added == 2021)
decline_pct <- round((1 - movie_latest$n / movie_peak$n) * 100)

p_big_idea <- ggplot(yearly_type |> filter(type == "Movie"),
  aes(x = year_added, y = n)) +
  geom_line(color = "#1565C0", linewidth = 2) +
  geom_point(color = "#1565C0", size = 3) +
  # Highlight the decline
  geom_segment(aes(x = movie_peak$year_added, xend = 2021,
    y = movie_peak$n, yend = movie_peak$n),
    linetype = "dotted", color = "#888") +
  annotate("text", x = 2020.5, y = (movie_peak$n + movie_latest$n) / 2,
    label = paste0(decline_pct, "% decline\nsince peak"),
    size = 4, fontface = "bold", color = "#E53935") +
  annotate("segment", x = 2020.5, xend = 2020.5,
    y = movie_peak$n - 10, yend = movie_latest$n + 10,
    arrow = arrow(ends = "both", length = unit(0.15, "cm")),
    color = "#E53935", linewidth = 0.8) +
  theme_minimal(base_size = 11) +
  theme(plot.title = element_text(face = "bold", size = 12)) +
  labs(title = paste0("Movie additions declined ", decline_pct,
    "% from peak — the content pipeline is contracting"),
       subtitle = "Netflix movie catalog growth has reversed; strategic pivot needed",
       x = NULL, y = "Movie Titles Added",
       caption = "Source: Netflix dataset")
ggsave("output/big_idea.png", p_big_idea, width = 9, height = 5, dpi = 300)

# ── 5. STORY ARC: THREE-SLIDE SEQUENCE ──────────────────
# Slide 1 — Context
p_context <- ggplot(nf |>
  filter(!is.na(date_added)) |>
  mutate(month = floor_date(date_added, "month")) |>
  count(month) |>
  mutate(cumulative = cumsum(n)),
  aes(x = month, y = cumulative)) +
  geom_area(fill = "#1565C0", alpha = 0.2) +
  geom_line(color = "#1565C0", linewidth = 1) +
  theme_minimal() +
  labs(title = "Netflix built a catalog of 8,800+ titles by 2021",
       subtitle = "Context: the platform grew rapidly over a decade",
       x = NULL, y = "Cumulative Titles")

# Slide 2 — Complication
p_complication <- p_explanatory +
  labs(title = "But movie additions have been declining since 2019",
       subtitle = "Complication: the largest content category is contracting")

# Slide 3 — Resolution
genre_growth <- nf_genres |>
  filter(!is.na(year_added), year_added %in% c(2017, 2021)) |>
  count(year_added, genre) |>
  pivot_wider(names_from = year_added, values_from = n, values_fill = 0) |>
  mutate(growth = (`2021` - `2017`) / pmax(`2017`, 1) * 100) |>
  slice_max(growth, n = 5)

p_resolution <- ggplot(genre_growth,
  aes(x = fct_reorder(genre, growth), y = growth)) +
  geom_col(fill = "#2E7D32", width = 0.5) +
  geom_text(aes(label = paste0("+", round(growth), "%")),
    hjust = -0.1, size = 3, fontface = "bold", color = "#2E7D32") +
  coord_flip(clip = "off") +
  theme_minimal() +
  labs(title = "International and niche genres are the growth engines",
       subtitle = "Resolution: pivot investment toward these categories",
       x = NULL, y = "Growth % (2017 → 2021)")

ggsave("output/story_arc_3slides.png",
  p_context / p_complication / p_resolution +
  plot_annotation(title = "Three-Slide Story Arc: Context → Complication → Resolution",
    tag_levels = "1"),
  width = 10, height = 14, dpi = 300)

# ── 6. DESCRIPTIVE vs DECLARATIVE TITLES ─────────────────
p_descriptive <- ggplot(yearly_type, aes(x = year_added, y = n, color = type)) +
  geom_line(linewidth = 1) + geom_point(size = 2) +
  scale_color_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  theme_minimal(base_size = 9) +
  labs(title = "Descriptive: 'Netflix Additions by Type'",
       subtitle = "(What the chart shows — not what it means)")

p_declarative <- ggplot(yearly_type, aes(x = year_added, y = n, color = type)) +
  geom_line(linewidth = 1) + geom_point(size = 2) +
  scale_color_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  theme_minimal(base_size = 9) +
  labs(title = "Declarative: 'Movie additions peaked in 2019; TV grew steadily'",
       subtitle = "(The title IS the finding — the chart is the evidence)")

ggsave("output/title_comparison.png",
  p_descriptive | p_declarative, width = 14, height = 5, dpi = 300)

cat("\n── All W07-M01 R outputs saved ──\n")
cat("Figures: exploratory_version, explanatory_version, exp_vs_expl,\n")
cat("  big_idea, story_arc_3slides, title_comparison\n")
